Do not check for known root datacenter names in vmware_guest

rpasche opened the pull request:
Just do not rely on known 'datacenter' names. For example,
on a german windows vCenter server, the root datacenter is
called 'Datencenter'. Hence, the folder path will always be wrong.

Fixes #29043

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
module_utils/vmware vmware_guest

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0 (devel e4c9ffa7e6) last updated 2017/09/25 08:02:01 (GMT +000)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible
  python version = 2.7.12 (default, Jun 29 2016, 08:57:23) [GCC 5.3.0]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->

Playbook given (cutted...)

```
      vmware_guest:
        hostname: "{{ vcenter }}"
        username: "{{ user }}"
        password: "{{ pass }}"
        validate_certs: no
        annotation: "Created by ansible"
        state: present
        resource_pool: Resources
        datacenter: TEST
        cluster: "TEST"
        name: "ansible_test1"
        folder: "ansible_test"
        guest_id: "debian7_64Guest"
```

Before
```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_I23I2j/ansible_module_vmware_guest.py", line 1503, in <module>
    main()
  File "/tmp/ansible_I23I2j/ansible_module_vmware_guest.py", line 1494, in main
    result = pyv.deploy_vm()
  File "/tmp/ansible_I23I2j/ansible_module_vmware_guest.py", line 1189, in deploy_vm
    dcpath = compile_folder_path_for_object(datacenter)
  File "/tmp/ansible_I23I2j/ansible_modlib.zip/ansible/module_utils/vmware.py", line 237, in compile_folder_path_for_object
AttributeError: 'NoneType' object has no attribute 'name'
```
This is because our root datacenter name is `Datencenter` and not `Datacenters`

After patch...vm is created correctly in the given folder (also cutted)

```
    "invocation": {
        "module_args": {
            "annotation": "Created by ansible",
            "cluster": "TEST",
            "customization": {},
            "customvalues": [],
            "datacenter": "TEST",
            "disk": [
                {
                    "autoselect_datastore": true,
                    "size_gb": 30,
                    "type": "thin"
                }
            ],
            "esxi_hostname": null,
            "folder": "/ansible_test",
            "is_template": false,
            "linked_clone": false,
            "name": "ansible_test1",
            "name_match": "first",
            ...
            ...
            "resource_pool": "Resources",

```

Akasurde commented:
@rpasche Could you please take a look at this failures - https://app.shippable.com/github/ansible/ansible/runs/38575/42/console ?

Try running tests using Python3.

rpasche commented:
Hm....with this patch, I come further than before, but it failes with

```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_xow62ioj/ansible_module_vmware_guest.py", line 1503, in <module>
    main()
  File "/tmp/ansible_xow62ioj/ansible_module_vmware_guest.py", line 1494, in main
    result = pyv.deploy_vm()
  File "/tmp/ansible_xow62ioj/ansible_module_vmware_guest.py", line 1228, in deploy_vm
    (datastore, datastore_name) = self.select_datastore(vm_obj)
  File "/tmp/ansible_xow62ioj/ansible_module_vmware_guest.py", line 1044, in select_datastore
    datastores = self.cache.get_all_objs(self.content, [vim.Datastore])
  File "/tmp/ansible_xow62ioj/ansible_module_vmware_guest.py", line 450, in get_all_objs
    for k, v in objects.items():
RuntimeError: dictionary changed size during iteration

fatal: [vciks -> localhost]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_xow62ioj/ansible_module_vmware_guest.py\", line 1503, in <module>\n    main()\n  File \"/tmp/ansible_xow62ioj/ansible_module_vmware_guest.py\", line 1494, in main\n    result = pyv.deploy_vm()\n  File \"/tmp/ansible_xow62ioj/ansible_module_vmware_guest.py\", line 1228, in deploy_vm\n    (datastore, datastore_name) = self.select_datastore(vm_obj)\n  File \"/tmp/ansible_xow62ioj/ansible_module_vmware_guest.py\", line 1044, in select_datastore\n    datastores = self.cache.get_all_objs(self.content, [vim.Datastore])\n  File \"/tmp/ansible_xow62ioj/ansible_module_vmware_guest.py\", line 450, in get_all_objs\n    for k, v in objects.items():\nRuntimeError: dictionary changed size during iteration\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}
```` 
But this also happens with the `devel` branch

```
ansible 2.5.0 (devel e4c9ffa7e6) last updated 2017/09/25 08:02:01 (GMT +000)
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible
  python version = 3.5.2 (default, Aug 18 2017, 17:48:00) [GCC 5.4.0 20160609]
```

rpasche commented:
Just some further information. Just tried a simple `vmware_guest_find` on my newly created vm. That works (with python 2.7 and python3.5)

```
---

- hosts: "{{ vc }}"
  connection: local
  gather_facts: no

  tasks:
    - name: find vm
      vmware_guest_find:
        datacenter: Stuttgart
        hostname: "{{ vc }}"
        username: "{{ user }}"
        password: "{{ pass }}"
        name: "ansible_test1"
        validate_certs: no
      delegate_to: localhost
```

Results

```
ansible-playbook 2.5.0 (find_datacenter_root_folder 703835d15d) last updated 2017/09/25 08:05:01 (GMT +000)
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible-playbook
  python version = 3.5.2 (default, Aug 18 2017, 17:48:00) [GCC 5.4.0 20160609]
No config file found; using defaults
Parsed /etc/ansible/hosts inventory source with ini plugin
...
...
ok: [vc -> localhost] => {
    "changed": false,
    "failed": false,
    "folders": [
        "/vm/ansible_test"
    ],
    "invocation": {
        "module_args": {
            "datacenter": "Stuttgart",
            "hostname": "{{ vc }}",
            "name": "ansible_test1",
            "password": "...",
            "username": "{{ user }}",
            "uuid": null,
            "validate_certs": false
        }
    }
}
```

rpasche commented:
@Akasurde Again further update. Changing

```
--- a/lib/ansible/modules/cloud/vmware/vmware_guest.py
+++ b/lib/ansible/modules/cloud/vmware/vmware_guest.py
@@ -447,10 +447,13 @@ class PyVmomiCache(object):
         if confine_to_datacenter:
             if hasattr(objects, 'items'):
                 # resource pools come back as a dictionary
+                # make temp copy
+                tmpobjs = objects.copy()
                 for k, v in objects.items():
                     parent_dc = self.get_parent_datacenter(k)
                     if parent_dc.name != self.dc_name:
-                        objects.pop(k, None)
+                        tmpobjs.pop(k, None)
+                objects = tmpobjs
             else:
                 # everything else should be a list
                 objects = [x for x in objects if self.get_parent_datacenter(x).name == self.dc_name]
````
fixes the 2nd error mentioned above when using `python3`

`RuntimeError: dictionary changed size during iteration`

and I can create the vm.

Full output

```
root@c15d3cc036e5:/ansible/playbooks# ansible-playbook -vvv test.yml
ansible-playbook 2.5.0 (fix_object_change_in_runtime 955107a39a) last updated 2017/09/25 13:38:48 (GMT +000)
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible-playbook
  python version = 3.5.2 (default, Aug 18 2017, 17:48:00) [GCC 5.4.0 20160609]
No config file found; using defaults
Parsed /etc/ansible/hosts inventory source with ini plugin

PLAYBOOK: test.yml *********************************************************************************************************
1 plays in test.yml

PLAY [vc] ***************************************************************************************************************
META: ran handlers

TASK [Create vm] ***********************************************************************************************************
task path: /ansible/playbooks/test.yml:8
Using module file /ansible/lib/ansible/modules/cloud/vmware/vmware_guest.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1506346910.1147048-178291982537145 `" && echo ansible-tmp-1506346910.1147048-178291982537145="` echo /root/.ansible/tmp/ansible-tmp-1506346910.1147048-178291982537145 `" ) && sleep 0'
<localhost> PUT /tmp/tmpzv9lt226 TO /root/.ansible/tmp/ansible-tmp-1506346910.1147048-178291982537145/vmware_guest.py
<localhost> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1506346910.1147048-178291982537145/ /root/.ansible/tmp/ansible-tmp-1506346910.1147048-178291982537145/vmware_guest.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python3 /root/.ansible/tmp/ansible-tmp-1506346910.1147048-178291982537145/vmware_guest.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1506346910.1147048-178291982537145/" > /dev/null 2>&1 && sleep 0'
changed: [vc -> localhost] => {
    "changed": true,
    "failed": false,
    "instance": {
        "annotation": "Created by ansible",
        "current_snapshot": null,
        "customvalues": {},
        "guest_tools_status": "guestToolsNotRunning",
        "guest_tools_version": null,
        "hw_eth0": {
            "addresstype": "assigned",
            "ipaddresses": null,
            "label": "Netzwerkadapter 1",
            "macaddress": "{{ mac }}",
            "macaddress_dash": "{{ mac-dash }}",
            "summary": "{{ pg }}"
        },
        "hw_guest_full_name": null,
        "hw_guest_id": null,
        "hw_interfaces": [
            "eth0"
        ],
        "hw_memtotal_mb": 1024,
        "hw_name": "ansible_test1",
        "hw_power_status": "poweredOff",
        "hw_processor_count": 2,
        "hw_product_uuid": "{{ uuid }}",
        "ipv4": null,
        "ipv6": null,
        "module_hw": true,
        "snapshots": []
    },
    "invocation": {
        "module_args": {
            "annotation": "Created by ansible",
            "cluster": "TEST",
            "customization": {},
            "customvalues": [],
            "datacenter": "Stuttgart",
            "disk": [
                {
                    "autoselect_datastore": true,
                    "size_gb": 30,
                    "type": "thin"
                }
            ],
            "esxi_hostname": null,
            "folder": "/ansible_test",
            "force": false,
            "guest_id": "debian7_64Guest",
            "hardware": {
                "memory_mb": 1024,
                "num_cpus": 2,
                "scsi": "lsilogic"
            },
            "hostname": "{{ vc }}",
            "is_template": false,
            "linked_clone": false,
            "name": "ansible_test1",
            "name_match": "first",
            "networks": [
                {
                    "device_type": "e1000",
                    "name": "{{ pg }}"
                }
            ],
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "resource_pool": "Resources",
            "snapshot_src": null,
            "state": "present",
            "template": null,
            "username": "{{ user }}",
            "uuid": null,
            "validate_certs": false,
            "wait_for_ip_address": false
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP *****************************************************************************************************************
vc                      : ok=1    changed=1    unreachable=0    failed=0
```

This has been run agains a german windows vCenter server 5.5

rpasche commented:
Further information. Because I'm pretty new to ansible, I did not know the `vcenter simulator`. Just to say....I'm still trying to rebuild this playbook with the vcenter simulator, but had no luck yet, using the default "infrastructure" within the simulator. Therefore I'm trying to build my own "infrastructure" within the simulator. If I further have time, I will try to install another vCenter appliance and try to setup a test infrastructure that also uses the structure `/<Folder>/<datacenter>/`

Akasurde commented on lib/ansible/module_utils/vmware.py:
```diff
@@ -234,7 +234,7 @@ def compile_folder_path_for_object(vobj):
     while hasattr(thisobj, 'parent'):
         thisobj = thisobj.parent
-        if thisobj.name == 'Datacenters':
+        if (not hasattr(thisobj, 'parent')) or (hasattr(thisobj, 'parent') and thisobj.parent is None):
```
How about 

```Python
    while True:
        try:
            thisobj = thisobj.parent
        except (IndexError, AttributeError) as err:
            break
        if isinstance(thisobj, vim.Folder):
            paths.append(thisobj.name)
    paths.reverse()
    return '/' + '/'.join(paths)
```

Akasurde replied:
This works in both Python 2 and 3. Here are few [tests](https://github.com/Akasurde/pyvmomi_examples/blob/master/compile_path_example.py#L54)

rpasche replied:
Looks pretty good to me. Will give it a try

Akasurde replied:
Cool. Let me know results for this.

mattclay commented:
CI has failed twice on this PR. I'm not sure if the issue is related or not. The latest failure is:

https://app.shippable.com/github/ansible/ansible/runs/38865/42/tests

I'll restart CI again to see if the issue goes away.

mattclay commented:
CI failed again with the same error. The failure appears to be related to changes made in this PR.

Akasurde commented:
@mattclay yes. The failures are due to changes in this PR. Thanks for help.

rpasche commented:
@mattclay @Akasurde 

How about this https://github.com/rpasche/pyvmomi_examples/blob/97d3897214515c31bdc5326572964a1e334ffa62/compile_path_example.py#L35

I did not yet add this changes to my PR yet (was very late yesterday)

@Akasurde Your suggested does not work for me. It goes one level too high and includes the root folder `Datacenters` (or on our german vcenter `Datencenter`) as part of the path, what is wrong. Also, the path is wrong.

Your code agains the vcsim

VCSIM inventory

```
bash-4.3# curl http://vcsim:5000/govc_find
[
  "/",
  "/F0",
  "/F0/DC0",
  "/F0/DC0/vm",
  "/F0/DC0/vm/F0",
  "/F0/DC0/vm/F0/DC0_H0_VM0",
  "/F0/DC0/vm/F0/DC0_H0_VM1",
  "/F0/DC0/vm/F0/DC0_C0_RP0_VM0",
  "/F0/DC0/vm/F0/DC0_C0_RP0_VM1",
  "/F0/DC0/vm/F0/DC0_C1_RP0_VM0",
  "/F0/DC0/vm/F0/DC0_C1_RP0_VM1",
  "/F0/DC0/host",
  "/F0/DC0/host/F0",
  "/F0/DC0/host/F0/DC0_H0",
  "/F0/DC0/host/F0/DC0_H0/DC0_H0",
  "/F0/DC0/host/F0/DC0_H0/Resources",
  "/F0/DC0/host/F0/DC0_C0",
  "/F0/DC0/host/F0/DC0_C0/DC0_C0_H0",
  "/F0/DC0/host/F0/DC0_C0/DC0_C0_H1",
  "/F0/DC0/host/F0/DC0_C0/DC0_C0_H2",
  "/F0/DC0/host/F0/DC0_C0/Resources",
  "/F0/DC0/host/F0/DC0_C0/Resources/DC0_C0_RP1",
  "/F0/DC0/host/F0/DC0_C1",
  "/F0/DC0/host/F0/DC0_C1/DC0_C1_H0",
  "/F0/DC0/host/F0/DC0_C1/DC0_C1_H1",
  "/F0/DC0/host/F0/DC0_C1/DC0_C1_H2",
  "/F0/DC0/host/F0/DC0_C1/Resources",
  "/F0/DC0/host/F0/DC0_C1/Resources/DC0_C1_RP1",
  "/F0/DC0/datastore",
  "/F0/DC0/datastore/DC0_POD0",
  "/F0/DC0/datastore/F0",
  "/F0/DC0/datastore/F0/LocalDS_0",
  "/F0/DC0/network",
  "/F0/DC0/network/VM Network",
  "/F0/DC0/network/F0",
  "/F0/DC0/network/F0/DVS0",
  "/F0/DC0/network/F0/DC0_DVPG0"
]
```
Your code gives

```
bash-4.3# python get_folder_vcsim
/Datacenters/F0
/Datacenters/F0/vm/F0
/Datacenters/F0/datastore/F0
/Datacenters/F0/network/F0
/Datacenters/F0/host/F0
```
The path for the datacenter is would be ok (if `/Datacenters` is removed), but it is wrong for the other objects because it includes the path the `datacenter`is in.

My version

```
bash-4.3# python get_folder_vcsim_rpasche
/F0
/vm/F0
/datastore/F0
/network/F0
/
```
When getting [here](https://github.com/ansible/ansible/blob/55739c1e46209dce06702586a0651e8b92a9d384/lib/ansible/modules/cloud/vmware/vmware_guest.py#L1204:L1216) the path sould be build correctly.

What I'm not sure is, if this is the correct folder "path" for a `resource pool`. Or if this is relevant for `resource_pool` to specify a `path`

Another difference between VCSIM and my "real" vcenter. When I'm on the root `datacenters` folder ("group-d1") and call `thisobj.parent` it results in `None` whereas it is an `IndexException` on VCSIM.

rpasche committed:
Do not check for known root datacenter names in vmware_guest

Just do not rely on known 'datacenter' names. For example,
on a german windows vCenter server, the root datacenter is
called 'Datencenter'. Hence, the folder path will always be wrong.

Fixes #29043

rpasche commented:
I `push --force` this branch (according to your guidelines). With this code, I can successfully create (and find) a new virtual machine on a german vCenter.

```
---

- hosts: "{{ vc }}"
  gather_facts: no
  connection: local

  tasks:
    - name: Create vm
      vmware_guest:
        hostname: "{{ vc }} "
        username: "{{ user }}"
        password: "{{ pass }}"
        validate_certs: no
        annotation: "Created by ansible"
        state: present
        resource_pool: Resources
        datacenter: Stuttgart
        cluster: "STG R730"
        name: "ansible_test3"
        folder: "ansible_test/ordner2"
        guest_id: "debian7_64Guest"
        hardware:
          num_cpus: 2
          memory_mb: 1024
          scsi: lsilogic
        disk:
          - size_gb: 30
            type: thin
            autoselect_datastore: yes
        networks:
          - name: "{{ pg }}"
            device_type: e1000
      delegate_to: localhost
```
Results in

```
changed: [vc -> localhost] => {
    "changed": true,
    "failed": false,
    "instance": {
        "annotation": "Created by ansible",
        "current_snapshot": null,
        "customvalues": {},
        "guest_tools_status": "guestToolsNotRunning",
        "guest_tools_version": null,
        "hw_eth0": {
            "addresstype": "assigned",
            "ipaddresses": null,
            "label": "Netzwerkadapter 1",
            "macaddress": "mac",
            "macaddress_dash": "mac",
            "summary": "pg"
        },
        "hw_guest_full_name": null,
        "hw_guest_id": null,
        "hw_interfaces": [
            "eth0"
        ],
        "hw_memtotal_mb": 1024,
        "hw_name": "ansible_test3",
        "hw_power_status": "poweredOff",
        "hw_processor_count": 2,
        "hw_product_uuid": "42211357-ce2b-e9ee-73a1-6f121314a86f",
        "ipv4": null,
        "ipv6": null,
        "module_hw": true,
        "snapshots": []
    },
    "invocation": {
        "module_args": {
            "annotation": "Created by ansible",
            "cluster": "STG R730",
            "customization": {},
            "customvalues": [],
            "datacenter": "Stuttgart",
            "disk": [
                {
                    "autoselect_datastore": true,
                    "size_gb": 30,
                    "type": "thin"
                }
            ],
            "esxi_hostname": null,
            "folder": "/ansible_test/ordner2",
            "force": false,
            "guest_id": "debian7_64Guest",
            "hardware": {
                "memory_mb": 1024,
                "num_cpus": 2,
                "scsi": "lsilogic"
            },
            "hostname": "vc",
            "is_template": false,
            "linked_clone": false,
            "name": "ansible_test3",
            "name_match": "first",
            "networks": [
                {
                    "device_type": "e1000",
                    "name": "pg"
                }
            ],
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "resource_pool": "Resources",
            "snapshot_src": null,
            "state": "present",
            "template": null,
            "username": "user",
            "uuid": null,
            "validate_certs": false,
            "wait_for_ip_address": false
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ******************************************************************************************************************
vc                      : ok=1    changed=1    unreachable=0    failed=0
```

Find

```
---

- hosts: "{{ vc }}"
  connection: local
  gather_facts: no

  tasks:
    - name: find vm
      vmware_guest_find:
        datacenter: Stuttgart
        hostname: "{{ vc }}"
        username: "{{ user }}"
        password: "{{ pass }}"
        name: "ansible_test3"
        validate_certs: no
      delegate_to: localhost
```

Results in

```
ok: [vciks -> localhost] => {
    "changed": false,
    "failed": false,
    "folders": [
        "/vm/ansible_test/ordner2"
    ],
    "invocation": {
        "module_args": {
            "datacenter": "Stuttgart",
            "hostname": "vc",
            "name": "ansible_test3",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "username": "user",
            "uuid": null,
            "validate_certs": false
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ******************************************************************************************************************
vci                      : ok=1    changed=0    unreachable=0    failed=0
```
I could not yet test to create a vm in a datacenter within a folder as I have no spare ESXi hosts yet in there.

jctanner commented:
Superseded by https://github.com/ansible/ansible/pull/31077
